include/revng/Pipeline/Pipe.h

```diff
@@ -0,0 +1,311 @@
+                const std::integer_sequence<size_t, S...> &) {
+  using namespace std;
+  (Pipe.*F)(Ctx,
+            getContainerFromName<decay_t<Args>>(Containers, ArgsNames[S])...);
+}
+
+template<typename PipeType, typename... Args, size_t... S>
+void invokeImpl(PipeType &Pipe,
+                void (PipeType::*F)(Args...),
+                ContainerSet &Containers,
+                const StringArrayRef &ArgsNames,
+                const std::integer_sequence<size_t, S...> &) {
+  using namespace std;
+  (Pipe.*F)(getContainerFromName<decay_t<Args>>(Containers, ArgsNames[S])...);
+}
+
+///
+/// Invokes the F member function on the Pipe Pipe passing as nth argument
+/// the container with the name equal to the nth element of ArgsNames
+///
+template<typename PipeType, typename... Args>
+void invokePipeFunction(Context &,
+                        PipeType &Pipe,
+                        void (PipeType::*F)(Args...),
+                        ContainerSet &Containers,
+                        const llvm::ArrayRef<std::string> &ArgsNames) {
+  constexpr auto
+    Indexes = std::make_integer_sequence<size_t, sizeof...(Args)>();
+  revng_assert(sizeof...(Args) == ArgsNames.size());
+  invokeImpl(Pipe, F, Containers, ArgsNames, Indexes);
+}
+
+template<typename PipeType, typename... Args>
+void invokePipeFunction(Context &Ctx,
+                        PipeType &Pipe,
+                        void (PipeType::*F)(Context &, Args...),
+                        ContainerSet &Containers,
+                        const llvm::ArrayRef<std::string> &ArgsNames) {
+  constexpr auto
+    Indexes = std::make_integer_sequence<size_t, sizeof...(Args)>();
+  revng_assert(sizeof...(Args) == ArgsNames.size());
+  invokeImpl(Ctx, Pipe, F, Containers, ArgsNames, Indexes);
+}
+
+template<typename PipeType, typename... Args>
+void invokePipeFunction(Context &Ctx,
+                        PipeType &Pipe,
+                        void (PipeType::*F)(const Context &, Args...),
+                        ContainerSet &Containers,
+                        const llvm::ArrayRef<std::string> &ArgsNames) {
+  constexpr auto
+    Indexes = std::make_integer_sequence<size_t, sizeof...(Args)>();
+  revng_assert(sizeof...(Args) == ArgsNames.size());
+  invokeImpl(Ctx, Pipe, F, Containers, ArgsNames, Indexes);
+}
+
+class PipeWrapperBaseImpl {
+public:
+  virtual void run(Context &Ctx, ContainerSet &Containers) = 0;
+  virtual ContainersStatus
+  getRequirements(const ContainersStatus &Target) const = 0;
+  virtual ContainersStatus deduceResults(ContainersStatus &Target) const = 0;
+  virtual ~PipeWrapperBaseImpl() = default;
+  virtual std::unique_ptr<PipeWrapperBaseImpl> clone() const = 0;
+  virtual std::vector<std::string> getRunningContainersNames() const = 0;
+  virtual std::string getName() const = 0;
+  virtual void dump(std::ostream &OS, size_t indents) const = 0;
+};
+
+template<typename T>
+concept Printable = requires(T a) {
+  { a.print(dbg, 0) };
```

Comment:
We usually use `template<typename S> void dump(S &)`.

Reply:
it's different here, there is dump for debug stuff and print for pretty print stuff.

Reply:
Are you sure they cannot be merged? What's an example output of print vs dump?